### PR TITLE
Fix intersection

### LINKED ISSUE #420: Weird level-deletion in `intersectionArrayBy`
https://github.com/haskell-unordered-containers/unordered-containers/blob/4da2c201a274f7f3306117b479c51aef7af61129/Data/HashMap/Internal.hs#L1862-L1865

I'm referring to line 1864 where, instead of returning a `BitmapIndexed` node, its only element is returned.

@oberblastmeister Could you clarify the reasoning behind this code?

I'm wondering whether this might mess with the shift/subkey-logic, effectively misplacing some elements, so they can't be found by lookups, deletions etc.

### PR DESCRIPTION
**Pull request by oberblastmeister**
Fixes #420

### CODE REVIEW

File: Data/HashMap/Internal.hs [hunk 1]
```diff
@@ -1859,9 +1859,10 @@ intersectionArrayBy f !b1 !b2 !ary1 !ary2
             testBit x = x .&. m /= 0
             b' = b .&. complement m
     (len, bFinal) <- go 0 0 0 bCombined bIntersect
+    l <- A.read mary 0
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:1862]: This read is only needed when `len == 1`. Can you move it into that branch?

**Comment by sjakobi** [comment]
@oberblastmeister Could you cherry-pick the regression test from #422 to this branch?

File: Data/HashMap/Internal.hs [hunk 2]
```diff
@@ -1861,7 +1861,11 @@ intersectionArrayBy f !b1 !b2 !ary1 !ary2
     (len, bFinal) <- go 0 0 0 bCombined bIntersect
     case len of
       0 -> pure Empty
-      1 -> A.read mary 0
+      1 -> do
+        l <- A.read mary 0
+        if isLeafOrCollision l
+          then pure l
+          else BitmapIndexed bFinal <$> (A.unsafeFreeze =<< A.shrink mary len)
```
Comment (treeowl) [inline Data/HashMap/Internal.hs:1868]: Wouldn't it be clearer to explicitly shrink to 1?

**Comment by sjakobi** [comment]
Thanks, @oberblastmeister. I'll finish this myself in #427.

**Pull request closed without merging**